pkg/connection/store/external/store.go

```diff
@@ -0,0 +1,147 @@
+)
+
+// SecretStore is an External Secret Store.
+type SecretStore struct {
+	client     ess.ExternalSecretStoreServiceClient
+	kubeClient client.Client
+	config     *v1.Config
+}
+
+// NewSecretStore returns a new External SecretStore.
+func NewSecretStore(ctx context.Context, kube client.Client, tlsConfig *tls.Config, cfg v1.SecretStoreConfig) (*SecretStore, error) {
+	if cfg.Plugin.Endpoint == nil {
+		return nil, errors.New("endpoint is not provided")
+	}
+
+	creds := credentials.NewTLS(tlsConfig)
+	conn, err := grpc.Dial(*(cfg.Plugin.Endpoint), grpc.WithTransportCredentials(creds))
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("cannot dial to the endpoint: %s", *cfg.Plugin.Endpoint))
+	}
+
+	cl := ess.NewExternalSecretStoreServiceClient(conn)
+
+	return &SecretStore{
+		kubeClient: kube,
+		client:     cl,
+		config:     cfg.Plugin.ConfigRef,
+	}, nil
+}
+
+// ReadKeyValues reads and returns key value pairs for a given Secret.
+func (ss *SecretStore) ReadKeyValues(ctx context.Context, n store.ScopedName, s *store.Secret) error {
+	sec := new(ess.Secret)
+	sec.ScopedName = filepath.Join(n.Scope, n.Name)
+
+	cfg := ss.getConfigReference()
+
+	res, err := ss.client.GetSecret(ctx, &ess.GetSecretRequest{Secret: sec, Config: cfg})
+	if err != nil {
+		return errors.Wrap(err, errGet)
+	}
+
+	s.ScopedName = n
+	s.Data = make(map[string][]byte, len(res.Secret.Data))
+	for d := range res.Secret.Data {
+		s.Data[d] = res.Secret.Data[d]
+	}
+	if res.Secret != nil && len(res.Secret.Metadata) != 0 {
+		s.Metadata = new(v1.ConnectionSecretMetadata)
+		s.Metadata.Labels = make(map[string]string, len(res.Secret.Metadata))
+		for k, v := range res.Secret.Metadata {
+			s.Metadata.Labels[k] = v
+		}
+	}
+
+	return nil
+}
+
+// WriteKeyValues writes key value pairs to a given Secret.
+func (ss *SecretStore) WriteKeyValues(ctx context.Context, s *store.Secret, wo ...store.WriteOption) (changed bool, err error) {
+	sec := new(ess.Secret)
+	sec.ScopedName = filepath.Join(s.Scope, s.Name)
+	sec.Data = make(map[string][]byte, len(s.Data))
+	for k, v := range s.Data {
+		sec.Data[k] = v
+	}
+
+	if s.Metadata != nil && len(s.Metadata.Labels) != 0 {
+		sec.Metadata = make(map[string]string, len(s.Metadata.Labels))
+		for k, v := range s.Metadata.Labels {
+			sec.Metadata[k] = v
+		}
+	}
+
+	cfg := ss.getConfigReference()
+
+	res, err := ss.client.ApplySecret(ctx, &ess.ApplySecretRequest{Secret: sec, Config: cfg})
+	if err != nil {
+		return false, errors.Wrap(err, errApply)
+	}
+
+	return res.Changed, nil
+}
+
+// DeleteKeyValues delete key value pairs from a given Secret.
+func (ss *SecretStore) DeleteKeyValues(ctx context.Context, s *store.Secret, do ...store.DeleteOption) error {
+	sec := new(ess.Secret)
+	sec.ScopedName = filepath.Join(s.Scope, s.Name)
+
+	cfg := ss.getConfigReference()
+
+	_, err := ss.client.DeleteKeys(ctx, &ess.DeleteKeysRequest{Secret: sec, Config: cfg})
+	if err != nil {
+		return errors.Wrap(err, errDelete)
+	}
+
+	return nil
```

Comment:
Nit: This could be simplified. `errors.Wrap` will return `nil` if the wrapped error is `nil`.


```suggestion
	_, err := ss.client.DeleteKeys(ctx, &ess.DeleteKeysRequest{Secret: sec, Config: cfg})
	return errors.Wrap(err, errDelete)
```